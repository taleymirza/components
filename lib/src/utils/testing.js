import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(mount(component)).toJSON()
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    const dest = computed[key] || (computed[key] = {})
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return div.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}
