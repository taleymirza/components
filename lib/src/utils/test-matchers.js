import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toSetDefaultTheme(Component) {
    const wrapper = mount(<Component />)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  }
})
