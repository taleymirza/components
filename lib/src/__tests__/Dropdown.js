import React from 'react'
import Dropdown from '../Dropdown'
import {render} from '../utils/testing'
import {COMMON} from '../constants'

describe('Dropdown', () => {
  it('matches the snapshots', () => {
    expect(render(<Dropdown>hi</Dropdown>)).toMatchSnapshot()
    expect(render(<Dropdown title="hi">hello!</Dropdown>)).toMatchSnapshot()
  })

  it('implements system props', () => {
    expect(Dropdown).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(Dropdown).toSetDefaultTheme()
  })
})

describe('Dropdown.Item', () => {
  it('matches the snapshots', () => {
    expect(render(<Dropdown.Item>hi</Dropdown.Item>)).toMatchSnapshot()
    expect(render(<Dropdown.Item>hello!</Dropdown.Item>)).toMatchSnapshot()
  })

  it('implements system props', () => {
    expect(Dropdown.Item).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(Dropdown.Item).toSetDefaultTheme()
  })
})

describe('Dropdown.Menu', () => {
  it('matches the snapshots', () => {
    expect(
      render(
        <Dropdown.Menu>
          <li>1</li>
          <li>2</li>
          <li>3</li>
        </Dropdown.Menu>
      )
    ).toMatchSnapshot()
    expect(
      render(
        <Dropdown.Menu>
          <li>1</li>
          <li>2</li>
          <li>3</li>
        </Dropdown.Menu>
      )
    ).toMatchSnapshot()
  })

  it('implements system props', () => {
    expect(Dropdown.Menu).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(Dropdown.Menu).toSetDefaultTheme()
  })
})
