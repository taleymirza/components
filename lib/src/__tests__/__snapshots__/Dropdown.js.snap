// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches the snapshots 1`] = `
.c2 {
  position: relative;
  display: inline-block;
  padding: 6px 12px;
  color: #24292e;
  background-color: #f6f8fa;
  background-image: linear-gradient(-180deg,#fafbfc 0%,rgb(239,243,246) 90%);
  font-size: 14px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-repeat: repeat-x;
  background-position: -1px -1px;
  background-size: 110% 110%;
  border: 1px solid rgba(27,31,35,0.2);
  border-radius: 0.25em;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 14px;
}

.c2:hover {
  background-color: rgb(230,235,241);
  background-image: linear-gradient(-180deg,rgb(239,243,246) 0%,rgb(230,235,241) 90%);
  background-position: -0.5em center;
  border-color: rgba(27,31,35,0.35);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-repeat: repeat-x;
}

.c2:active {
  background-color: rgb(233,236,239);
  background-image: none;
  box-shadow: rgba(27,31,35,0.15) 0px 0.15em 0.3em inset;
  border-color: rgba(27,31,35,0.2);
}

.c2.disabled {
  color: rgba(36,41,46,0.4)!important;
  background-color: #f6f8fa!important;
  background-image: none!important;
  border-color: rgba(27,31,35,0.20)!important;
  box-shadow: none!important;
  cursor: default;
}

.c2:focus {
  outline: none;
  box-shadow: rgba(3,102,214,0.3) 0px 0px 0px 0.2em;
}

.c2.grouped {
  position: relative;
  border-right-width: 0;
  border-radius: 0;
}

.c2.grouped:first-child {
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.c2.grouped:last-child {
  border-right-width: 1px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.c2.grouped:focus,
.c2.grouped:active,
.c2.grouped:hover {
  border-right-width: 1px;
}

.c2.grouped:focus + .grouped,
.c2.grouped:active + .grouped,
.c2.grouped:hover + .grouped {
  border-left-width: 0;
}

.c2:focus,
.c2:active {
  z-index: 1;
}

.c1 > summary {
  list-style: none;
}

.c1 > summary::-webkit-details-marker {
  display: none;
}

.c0 {
  position: relative;
  display: inline-block;
}

.c3 {
  border: 4px solid transparent;
  border-top-color: currentcolor;
  content: '';
  display: inline-block;
  height: 0;
  vertical-align: middle;
  width: 0;
}

<details
  className="c0 c1"
  open={false}
>
  <summary
    aria-haspopup="true"
    className="c0 c2"
    onClick={[Function]}
  >
    <div
      className="c3"
    />
  </summary>
  hi
</details>
`;

exports[`Dropdown matches the snapshots 2`] = `
.c2 {
  position: relative;
  display: inline-block;
  padding: 6px 12px;
  color: #24292e;
  background-color: #f6f8fa;
  background-image: linear-gradient(-180deg,#fafbfc 0%,rgb(239,243,246) 90%);
  font-size: 14px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-repeat: repeat-x;
  background-position: -1px -1px;
  background-size: 110% 110%;
  border: 1px solid rgba(27,31,35,0.2);
  border-radius: 0.25em;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 14px;
}

.c2:hover {
  background-color: rgb(230,235,241);
  background-image: linear-gradient(-180deg,rgb(239,243,246) 0%,rgb(230,235,241) 90%);
  background-position: -0.5em center;
  border-color: rgba(27,31,35,0.35);
  -webkit-text-decoration: none;
  text-decoration: none;
  background-repeat: repeat-x;
}

.c2:active {
  background-color: rgb(233,236,239);
  background-image: none;
  box-shadow: rgba(27,31,35,0.15) 0px 0.15em 0.3em inset;
  border-color: rgba(27,31,35,0.2);
}

.c2.disabled {
  color: rgba(36,41,46,0.4)!important;
  background-color: #f6f8fa!important;
  background-image: none!important;
  border-color: rgba(27,31,35,0.20)!important;
  box-shadow: none!important;
  cursor: default;
}

.c2:focus {
  outline: none;
  box-shadow: rgba(3,102,214,0.3) 0px 0px 0px 0.2em;
}

.c2.grouped {
  position: relative;
  border-right-width: 0;
  border-radius: 0;
}

.c2.grouped:first-child {
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.c2.grouped:last-child {
  border-right-width: 1px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.c2.grouped:focus,
.c2.grouped:active,
.c2.grouped:hover {
  border-right-width: 1px;
}

.c2.grouped:focus + .grouped,
.c2.grouped:active + .grouped,
.c2.grouped:hover + .grouped {
  border-left-width: 0;
}

.c2:focus,
.c2:active {
  z-index: 1;
}

.c1 > summary {
  list-style: none;
}

.c1 > summary::-webkit-details-marker {
  display: none;
}

.c0 {
  position: relative;
  display: inline-block;
}

.c3 {
  border: 4px solid transparent;
  border-top-color: currentcolor;
  content: '';
  display: inline-block;
  height: 0;
  vertical-align: middle;
  width: 0;
}

<details
  className="c0 c1"
  open={false}
>
  <summary
    aria-haspopup="true"
    className="c0 c2"
    onClick={[Function]}
  >
    hi
    <div
      className="c3"
    />
  </summary>
  hello!
</details>
`;

exports[`Dropdown.Item matches the snapshots 1`] = `
.c0 {
  display: block;
  padding: 4px 10px 4px 15px;
  overflow: hidden;
  color: $get('colors.gray.9');
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0:focus {
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #0366d6;
}

.c0:hover {
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #0366d6;
  outline: none;
}

<li
  className="c0"
>
  hi
</li>
`;

exports[`Dropdown.Item matches the snapshots 2`] = `
.c0 {
  display: block;
  padding: 4px 10px 4px 15px;
  overflow: hidden;
  color: $get('colors.gray.9');
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0:focus {
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #0366d6;
}

.c0:hover {
  color: #fff;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #0366d6;
  outline: none;
}

<li
  className="c0"
>
  hello!
</li>
`;

exports[`Dropdown.Menu matches the snapshots 1`] = `
.c0 {
  background-clip: padding-box;
  background-color: #fff;
  border: 1px solid rgba(27,31,35,0.15);
  border-radius: 4px;
  box-shadow: 0 3px 12px rgba(27,31,35,0.15);
  left: 0;
  list-style: none;
  margin-top: 2px;
  padding: 5px 0 5px 0 !important;
  position: absolute;
  top: 100%;
  width: 160px;
  z-index: 100;
  right: 0;
  left: auto;
}

.c0::before {
  position: absolute;
  display: inline-block;
  content: '';
}

.c0::after {
  position: absolute;
  display: inline-block;
  content: '';
}

.c0::before {
  border: 8px solid transparent;
  border-bottom-color: rgba(27,31,35,0.15);
}

.c0::after {
  border: 7px solid transparent;
  border-bottom-color: #fff;
}

.c0 > ul {
  list-style: none;
}

.c0::before {
  top: -16px;
  right: 9px;
  left: auto;
}

.c0::after {
  top: -14px;
  right: 10px;
  left: auto;
}

<ul
  className="c0"
  direction="sw"
>
  <li>
    1
  </li>
  <li>
    2
  </li>
  <li>
    3
  </li>
</ul>
`;

exports[`Dropdown.Menu matches the snapshots 2`] = `
.c0 {
  background-clip: padding-box;
  background-color: #fff;
  border: 1px solid rgba(27,31,35,0.15);
  border-radius: 4px;
  box-shadow: 0 3px 12px rgba(27,31,35,0.15);
  left: 0;
  list-style: none;
  margin-top: 2px;
  padding: 5px 0 5px 0 !important;
  position: absolute;
  top: 100%;
  width: 160px;
  z-index: 100;
  right: 0;
  left: auto;
}

.c0::before {
  position: absolute;
  display: inline-block;
  content: '';
}

.c0::after {
  position: absolute;
  display: inline-block;
  content: '';
}

.c0::before {
  border: 8px solid transparent;
  border-bottom-color: rgba(27,31,35,0.15);
}

.c0::after {
  border: 7px solid transparent;
  border-bottom-color: #fff;
}

.c0 > ul {
  list-style: none;
}

.c0::before {
  top: -16px;
  right: 9px;
  left: auto;
}

.c0::after {
  top: -14px;
  right: 10px;
  left: auto;
}

<ul
  className="c0"
  direction="sw"
>
  <li>
    1
  </li>
  <li>
    2
  </li>
  <li>
    3
  </li>
</ul>
`;
